Update module/model init schemes to match Llama2

daviswer opened the pull request:
Update the model init scheme to match Llama2, in anticipation for trialing the more principled auto-scaling scheme in the pretrain repo. Specifically:

- Update initialization scales to 0.02 across the board
- Layers no longer auto-initialize
- Change `reset_params` to `reset_parameters` to match PyTorch APIs
- Model no longer auto-initializes, as this is faster and the intention appears to be to load from checkpoint in most cases for this repo. `Reset_parameters()` should be used as a literal reset. 
- Update Llama `reset_parameters()` to call the `reset_parameters()` of sublayers
- Update Roberta and GPTBigCode to do the same (adaptive scaling of embedding/head layers is preserved from old init scheme, this can be revisited later)
- Remove `gain` parameters from relevant ff/attn layers. This is unused in the original init scheme, and can be utilized outside the model for the trial init scheme in the pretrain repo. 
- Update the `test_load` CI test to construct the initial model through `get_model` instead of `_get_model_instance`. This initializes the model properly before saving/loading/comparing the state dict, preventing comparison of non-deterministic NaN values in the uninitialized model.

nairbv commented on fms/models/llama.py:
```diff
@@ -218,8 +218,6 @@ def __init__(
         if self.config.p_dropout:
             self.dropout = nn.Dropout(self.config.p_dropout)
 
-        self.reset_params()
```
it probably does make sense to remove this, but I think the existing training script in this repo expects model instances it gets to be randomly initialized. we could can add a reset_params() call to that code later.

Then again, it's not defined in any super class, so calling code has to hope to find it. PyTorch modules use the convention `reset_parameters` instead of reset_params, should we change to that for the method name, to at least match pytorch?

daviswer replied:
Yeah that's a good idea, I'll do that. The downside is that I can no longer use `hasattr(m, "reset_parameters")` to check the submodules, but arguably I should be listing those submodules out explicitly anyways.

nairbv replied:
what is it that breaks about `hasattr(m, "reset_parameters")`?

I don't currently see any references to `reset_parameters` in fms or `pretrain` repos.

nairbv replied:
related change: https://github.com/foundation-model-stack/foundation-model-stack/pull/179

daviswer replied:
What breaks is that earlier I had been relying on the existence of a `reset_params` method to know when to call it. If every module has a `reset_parameters`, then I can't just call it whenever it occurs, as I could end up custom-initializing at the layer level, then re-initializing at the sublayer level with the wrong init scheme. But again NBD, just means I need to list out the custom modules explicitly in the init scheme

nairbv commented on fms/models/llama.py:
```diff
@@ -228,11 +226,10 @@ def from_config(cls, config: LLaMAConfig) -> "LLaMA":
-        self.shared.head.weight.data.normal_(
-            0, 1 / math.sqrt(math.sqrt(self.width * self.shared.vocab_size))
-        )
+        # Call reset_params for relevant sub-layers
```
does llama tie weights? i.e. should we avoid reseting parameters for `self.shared.head.weight` 2x?

daviswer replied:
The `reset_params` method of the embedding layer should take care of initializing the head, if not tying, or tying it, if not

nairbv replied:
ah for some reason I was thinking there might be two references to it, but of course it will only be present once in `modules()`

nairbv commented on fms/modules/attention.py:
```diff
@@ -79,25 +78,13 @@ def __init__(
-            )  # Using explicit terms instead of numel to account for eventual MQA addition
-        if self.use_bias:
-            for layer in ["query", "key", "value", "dense"]:
+        layers = ["query", "key", "value", "dense"]
```
instead of relying on the names of the attributes, is there some cleaner way to do this like self.modules() and filtering on isinstance nn.Linear?

raghukiran1224 commented:
I would just use the init as what worked for us, can we not have multiple of these? Let's just focus on what worked...

nairbv commented:
@JRosenkranz tests in this PR like `test_hf_model_compile_no_graph_breaks` fail, but all that has changed is the model init code. Could these failures somehow be related to signature changes or? Davis might need help on the command to regenerate signatures for the expect tests if so. I don't see anything in the output about regenerating the expects so maybe it's something else? If this is a signatures issue then we should find a way to better reflect it in the test output.

JRosenkranz commented:
> @JRosenkranz tests in this PR like `test_hf_model_compile_no_graph_breaks` fail, but all that has changed is the model init code. Could these failures somehow be related to signature changes or? Davis might need help on the command to regenerate signatures for the expect tests if so. I don't see anything in the output about regenerating the expects so maybe it's something else? If this is a signatures issue then we should find a way to better reflect it in the test output.

I would need to look into this, I don't think its related to expects test as its not checking for correctness in this test, only checking for model compiling with no graph breaks.

daviswer commented:
I've updated the reset function to match PyTorch's conventions, and Llama reset should now call these sub-resets correctly. Updated description to reflect this

lchu-ibm commented:
Can someone take a look at the check failure and see if it is related to this PR?  If not, we should land this PR.

lchu-ibm commented:
@daviswer I merged the main and re-triggered the test and it still failed.

@JRosenkranz the error seems unrelated and the same error is occurring in all other PRs except yours. Can you take a look at the test failure and provide some insights?

JRosenkranz commented:
> @JRosenkranz the error seems unrelated and the same error is occurring in all other PRs except yours. Can you take a look at the test failure and provide some insights?

The issue is that in llama we used to call reset_params by default and it has now been removed in this branch. One of the reset_params lines calls 

```python
self.rot_emb.compute_freqs_cis(
     self.shared.emb.weight.device, self.config.max_expected_seq_len
)
```

If you move the following lines into `__init__` (right below the rot_emb), this issue should go away:

```python
        if isinstance(self.distributed_strategy, UniformModelParallelStrategy):
            for dev_idx in set(self.distributed_strategy.layer_to_device):
                self.rot_emb.compute_freqs_cis(
                    torch.device("cuda", dev_idx), self.config.max_expected_seq_len
                )
        else:
            self.rot_emb.compute_freqs_cis(
                self.shared.emb.weight.device, self.config.max_expected_seq_len
            )
```

lchu-ibm commented:
@JRosenkranz Thanks!

@daviswer can you make the change?

daviswer commented:
Change made, now it looks like it's failing a checkpoint loading closeness test

daviswer commented:
I think the test is failing because this branch removes auto-initialization, so the tensors are default filled with random junk. We declare a fresh model, save the checkpoint, load it back, then check equivalence, but we end up comparing NaN to NaN. We need to either add a `reset_parameters()` call to the test, or add special handling for NaN/inf/neginf values (or add back model auto-init)

daviswer commented:
Figured out the issue - the test was constructing the model through `_get_model_instance` instead of `get_model`, which initializes the model properly. All tests passing!

nairbv commented on fms/models/llama.py:
```diff
@@ -195,6 +195,16 @@ def __init__(
             max_seq_len=self.config.max_expected_seq_len,
         )
+        # RoPE init
+        if isinstance(self.distributed_strategy, UniformModelParallelStrategy):
```
if we do this in the constructor I don't think it needs to be in the reset_parameters code too

daviswer replied:
Good point, I removed it from reset_parameters (as we should be calling this all the time)

nairbv commented on fms/models/llama.py:
```diff
@@ -218,21 +228,23 @@ def __init__(
+                isinstance(m, MultiHeadAttention)
+                or isinstance(m, WordEmbedding)
+                or isinstance(m, GatedLinearUnit)
+                or isinstance(m, LayerNormParameterized)
```
would it be better to just `hasattr(m, 'reset_parameters')` instead of listing them? We could end up adding more variants of these modules.

Also, for consistency, I think we need to update the other models too?

daviswer replied:
Unfortunately we can't just use `hasattr(n, 'reset_parameters')` because PyTorch submodules (including `nn.Linear`) also have a `reset_parameters`. So we could end up calling our initializer, then accidentally overriding it with the PyTorch defaults for the submodules. 

We should update the other models, I'll propagate these changes to Roberta/GPTBigCode (though it's not fully clear to what degree this init scheme is suited toward these models in practice)

daviswer replied:
Updated other model `reset_parameters` and updated description to reflect this

nairbv replied:
ah, I think to do it the way I was suggesting it would need to be `.children()` instead of .`modules()` and then there'd need to be a similar `reset_parameters` method on LLaMABlock. Then it would only trickle through on the explicitly added modules and not flow into underlying linear modules. I guess this is fine too though.

daviswer replied:
Oh I see, yeah makes sense
